pilot/pkg/xds/endpoint_builder.go

```diff
@@ -124,30 +124,31 @@ func (b EndpointBuilder) DestinationRule() *networkingapi.DestinationRule {
 
 // Key provides the eds cache key and should include any information that could change the way endpoints are generated.
 func (b EndpointBuilder) Key() string {
-	params := []string{
-		b.clusterName,
-		string(b.network),
-		string(b.clusterID),
-		strconv.FormatBool(b.clusterLocal),
-		util.LocalityToString(b.locality),
-		b.tunnelType.ToString(),
-	}
+	hash := md5.New()
+	hash.Write([]byte(b.clusterName))
+	hash.Write([]byte(b.network))
+	hash.Write([]byte(b.clusterID))
+	hash.Write([]byte(strconv.FormatBool(b.clusterLocal)))
+	hash.Write([]byte(util.LocalityToString(b.locality)))
+	hash.Write([]byte(b.tunnelType.ToString()))
+
 	if b.push != nil && b.push.AuthnPolicies != nil {
-		params = append(params, b.push.AuthnPolicies.GetVersion())
+		hash.Write([]byte(b.push.AuthnPolicies.GetVersion()))
 	}
 	if b.destinationRule != nil {
-		params = append(params, b.destinationRule.Name+"/"+b.destinationRule.Namespace)
+		hash.Write([]byte(b.destinationRule.Name))
+		hash.Write([]byte{'/'})
```

Comment:
nit: not sure we really need the `/`. I guess we didn't really before either.

In the original form long ago we sort of did, before we did a hash.

Reply:
it could theoretically reduce chances of collision, when two different names and namespaces have the same result when appending, though I don't know how often this could happen.
for example: `name + ns = nam + ens`

let me know if I should remove them or keep them

Reply:
ah good point.

I wonder if we need divider for everything then. In the original one we did `strings.Join(params, "~")` to avoid. If write(ab) is the same as `write(a); write(b)` then I think its actually unsafe the way we do things now (with and without this change)?

Reply:
Yup, `write(ab)` is the same as `write(a); write(b)`, i tested it, if we previously joined with `~` then we should write `~` before every other write

Reply:
I was thinking a little bit more about this, the chances of collision of all the entities we use in the cache key is probably very low, right? so maybe the dividers are not really that necessary?

On another hand, adding them is very easy, and could prevent some headaches in case we do get collisions.

Reply:
If there is no cost/ very low cost of adding them - I would just go ahead and add